From 83de069511deb2345695d3b29f48f9dab7d9debf Mon Sep 17 00:00:00 2001
From: Alaric Senat <dev.asenat@posteo.net>
Date: Mon, 27 May 2024 13:55:44 +0200
Subject: [PATCH 53/54] upnp: enforce HTTP(S) as the only supported protocols

In preparation for the next commit, explicitly refuse any protocols
that are not supported by libpupnp. It's unclear if it's needed and
pupnp probably already perform this check, but the next commit rely
heavily on the scheme matching "http" or "https" only.

(cherry picked from commit 44de051a61475a899f1bd71a3db0e737a481b41f)
---
 modules/services_discovery/upnp.cpp | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/modules/services_discovery/upnp.cpp b/modules/services_discovery/upnp.cpp
index fc5d93ca2b..bb131dd7e3 100644
--- a/modules/services_discovery/upnp.cpp
+++ b/modules/services_discovery/upnp.cpp
@@ -350,6 +350,17 @@ bool MediaServerList::addServer( MediaServerDesc* desc )
             free( psz_playlist_option );
         }
     } else {
+        const auto loc_starts_with = [desc](const char *s) {
+            return desc->location.compare(0, strlen(s), s) == 0;
+        };
+
+        if ( !loc_starts_with("http://") && !loc_starts_with("https://") )
+        {
+            msg_Warn( m_sd, "Unexpected underlying protocol, the UPNP module "
+                      "fully supports HTTP and has partial support for HTTPS" );
+            return false;
+        } 
+
         std::string mrl = std::string("upnp://") + desc->location;
 
         // Forge a root object ID in the MRL, this is used in the dir access.
-- 
2.39.3 (Apple Git-146)

